Stratis.Bitcoin.Features.Consensus/ConsensusLoop.cs

```diff
@@ -62,78 +164,188 @@ public void Initialize()
+        /// </summary>
+        public void Stop()
         {
-            while (true)
+            this.asyncLoop?.Dispose();
+        }
+
+        /// <summary>
+        /// A puller method that will continuously loop and ask for the next block  in the chain from peers.
+        /// The block will then be passed to the consensus validation. 
+        /// </summary>
+        /// <param name="cancellationToken">A cancellation token that will stop the loop.</param>
+        /// <returns>The task the loop will be running under.</returns>
+        private Task PullerLoop(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
             {
-                yield return this.ExecuteNextBlock(cancellationToken);
+                try
+                {
+                    BlockItem item = new BlockItem();
+
+                    using (new StopwatchDisposable(o => this.Validator.PerformanceCounter.AddBlockFetchingTime(o)))
+                    {
+                        item.Block = this.Puller.NextBlock(cancellationToken);
+                    }
+
+                    this.logger.LogTrace("Block received from puller.");
+                    this.AcceptBlock(item);
+
+                }
+                catch (Exception ex)
+                {
+                    if (ex is OperationCanceledException)
+                    {
+                        if (this.nodeLifetime.ApplicationStopping.IsCancellationRequested)
+                            return Task.FromException(ex);
+                    }
+
+                    // TODO Need to revisit unhandled exceptions in a way that any process can signal an exception has been
+                    // thrown so that the node and all the disposables can stop gracefully.
+                    this.logger.LogCritical(new EventId(0), ex, "Consensus loop at Tip:{0} unhandled exception {1}", this.Tip, ex.ToString());
+                    NLog.LogManager.Flush();
+                    throw;
+                }
             }
+
+            return Task.CompletedTask;
         }
 
-        public BlockResult ExecuteNextBlock(CancellationToken cancellationToken)
+        /// <summary>
+        /// A method that will accept a new block to the node.
+        /// The block will be validated and the <see cref="CoinView"/> db will be updated.
+        /// If its a new block that extends the chain then the <see cref="ConcurrentChain.Tip"/> will be set.
+        /// </summary>
+        /// <param name="item">The block to validate</param>
+        public void AcceptBlock(BlockItem item)
         {
             this.logger.LogTrace("()");
 
-            BlockResult result = new BlockResult();
-            try
+            lock (this.consensusLock)
             {
-                using (new StopwatchDisposable(o => this.Validator.PerformanceCounter.AddBlockFetchingTime(o)))
+                if (item.Block == null)
```

Comment:
are you sure this belongs here? if so, please make sure the method's param comment for `item` specifically mentions this specific configuration

however, i'm not sure why this is not in the puller loop, it could as well use this lock, you would just need to change its comment, which currently says that it is only used in accept block, instead it should say which shared structures it protects

Reply:
Hmmm thats a good point, this reorg can go in the puller loop, perhaps its even better there as this is reorg logic that is very much connected to the puller (block = null is a puller logic)

Ok yeah I will move it.

Reply:
Actually I did not move this yet to the puller loop as i am not sure how the async code will work.

Reply:
The issue I have here is the lock, and that depends on how we implement async.
I might decide to add a second loop for consensus that will be async and not require a lock at all then the rewind will have to go back to the AccpetBlock method.
However if we use the AsyncLock it might be find to move the code tot he puller loop.

Reply:
the last version i was testing did have this in the puller loop already
all that is needed there is just to replace the lock with AsyncLock and it should be fine